crates/diagnostic/src/lib.rs

```diff
@@ -505,6 +505,42 @@ impl Diagnostic {
             }
         }
     }
+
+    /// hir_tyのチェックエラー情報を元に診断情報を作成します。
+    pub fn from_hir_ty_check_error(
+        db: &base_db::SalsaDatabase,
+        _hir_file: hir::HirFile,
+        _file_db: &hir::HirFileDatabase,
+        source_map: &hir::HirFileSourceMap,
+        error: &hir_ty::TypeCheckError,
+    ) -> Self {
+        match error {
+            hir_ty::TypeCheckError::UnresolvedType { expr } => {
+                let text_range = expr.text_range(db, source_map);
+                Diagnostic {
+                    file: source_map.file(db),
+                    title: "Unresolved type".to_string(),
+                    head_offset: text_range.start().into(),
+                    messages: vec![Message {
+                        message: "cannot resolve type".to_string(),
+                        range: text_range,
+                    }],
+                }
+            }
+            hir_ty::TypeCheckError::ImmutableReassignment { expr } => {
+                let text_range = expr.text_range(db, source_map);
+                Diagnostic {
+                    file: source_map.file(db),
+                    title: "Immutable reassignment".to_string(),
+                    head_offset: text_range.start().into(),
+                    messages: vec![Message {
+                        message: "cannot reassign to immutable variable".to_string(),
+                        range: text_range,
+                    }],
+                }
+            }
+        }
+    }
```

Comment:
**Performance Issue**: The new function 'from_hir_ty_check_error' introduces potential performance issues related to error handling. The function performs type checks and creates diagnostic messages, which could be optimized to reduce computational overhead and improve response times in scenarios with high error rates. <br> **Fix**: Optimize the error handling by implementing more efficient data structures or algorithms for error processing. Consider lazy evaluation of error messages or caching common error results to reduce the computational cost. <br> **Code Suggestion**: 
 ```
 +            hir_ty::TypeCheckError::UnresolvedType { expr } => {
 +                let text_range = expr.text_range(db, source_map);
 +                // Consider caching common unresolved types to avoid recomputation
 +                Diagnostic {
 +                    file: source_map.file(db),
 +                    title: "Unresolved type".to_string(),
 +                    head_offset: text_range.start().into(),
 +                    messages: vec![Message {
 +                        message: "cannot resolve type".to_string(),
 +                        range: text_range,
 +                    }],
 +                }
 +            }
 ```